Rework I2C receiver to handle transactions of all lengths

WillB97 opened the pull request:
- Corrects some errors in the single byte read implementation to match RM0008 26.3.3 (Figure 277)
- Reworks the read > 2 bytes code to be functional and match figure 275
- Remove redundent NACK checking where the slave cannot NACK

Fixes #14.

WillB97 committed:
Rework I2C receiver to handle transactions of all lengths

richardbarlow reviewed (commented):
Sorry, I haven't done any STM32 stuff at this low level for many years now so I'm probably not the best person to review it. When I'm doing STM32 work these days I typically use the HAL or LL library from ST, which deals with all of these intricacies.

When we started the SR STM32 firmware the ST library licence was very poor, hence using libopencm3. However, since then ST have sorted the licence situation out and it is now BSD 3-clause. I would highly recommend using ST's libraries where possible, otherwise you'll forever be chasing niche bugs in how you configure peripherals rather than spending time on the application specific stuff that really matters. I'm sure @rgilton could give you a couple of pointers on how to sensibly integrate it, if you're interested.

WillB97 committed:
i2c_set_clock_frequency does not change the PCLK1 frequency.
It sets the frequency used to calculate slew rates
i2c_set_speed calls this internally

WillB97 commented:
That's certainly a good point and is already front of mind for the next kit design. But since this kit is approaching EOL I just want to do make the required changes to allow it to work for next years competition so we can put it down tidily. If we continue to use this further I definitely think it's worth pursuing using a HAL.
